pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,568 @@
+	}).Sub(float64(numBuilds))
+
+	b.reportStatus(q, false)
+
+	return numBuilds
+}
+
+// numFolded returns the number of folded builds
+func (b *build) numFolded() int {
+	return len(b.notificationChannels)
+}
+
+func newBuild(b Builder, uuid string) *build {
+	return &build{
+		builder:              b,
+		uuid:                 uuid,
+		notificationChannels: []BuildNotification{},
+		internalNotification: make(BuildNotification, 1),
+	}
+}
+
+// buildStatus is the current status of a build
+type buildStatus struct {
+	builder Builder
+
+	// curentBuild is the building information of any running build
+	currentBuild *build
+
+	// nextBuild is the building information of the next build for this
+	// UUID
+	nextBuild *build
+}
+
+func (q *BuildQueue) enqueueBuild(b Builder) (BuildNotification, bool) {
+	build, enqueued := q.serializeBuild(b)
+	notify := build.createAndAddNotificationChannel()
+
+	metrics.BuildQueueEntries.With(map[string]string{
+		metrics.LabelBuildQueueName:  q.name,
+		metrics.LabelBuildQueueState: metrics.BuildQueueWaiting,
+	}).Inc()
+
+	b.BuildQueued()
+
+	return notify, enqueued
+}
+
+func (q *BuildQueue) serializeBuild(b Builder) (*build, bool) {
+	uuid := b.GetUUID()
+	build := newBuild(b, uuid)
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	if buildStatus, ok := q.buildStatus[uuid]; ok {
+		// If the builder is currently being built, prepare the next
+		// build and store it in the nextBuild field. When the current
+		// build finishes, the next build will automatically be added
+		// to the workerBuildQueue
+		if buildStatus.currentBuild != nil && buildStatus.nextBuild == nil {
+			buildStatus.nextBuild = build
+			return build, false
+		}
+
+		// The builder is already in the queue but not being built, the
+		// build request will be fulfilled when the queued build is
+		// executed. The caller can be notified to skip the build.
+		return buildStatus.nextBuild, false
+	}
+
+	q.buildStatus[uuid] = &buildStatus{
+		builder:   b,
+		nextBuild: build,
+	}
+
+	return build, true
+}
+
+// Enqueue schedules Builder for building. A channel is returned to provide the
+// ability to wait for the build to complete and check for success or failure.
+//
+// A regular build fullfils the following build guarantees:
+// * The same Buildable, i.e. the same UUID, will not be built in parallel on
+//   multiple workers.
+// * The order in which regular builds are enqueued is maintained in the build
+//   order across different Buildables. If the same Buildable is enqueued
+//   multiple times, the buils are automatically folded together until the
+//   build is executed. This means in practise that a specific Buildable can only
+//   be in the build queue once.
+// * In the presence of preemptive, exclusive builds, regular builds get a fair
+//   chance to run but preemptive builds always preempt any queued, not yet
+//   running regular build.
+func (q *BuildQueue) Enqueue(b Builder) BuildNotification {
```

Comment:
To provide more safety, I would return a read-only channel here, instead of the read-write `BuildNotification`.